contracts/src/v0.8/dev/automation/upkeeps/MercuryRegistry.sol

```diff
@@ -0,0 +1,280 @@
+|   - Optimize gas consumption.                                                                                       |
+-+---------------------------------------------------------------------------------------------------------------------*/
+contract MercuryRegistry is ConfirmedOwner, AutomationCompatibleInterface, FeedLookupCompatibleInterface {
+  error DuplicateFeed(string feedId);
+  error FeedNotActive(string feedId);
+  error StaleReport(string feedId, uint32 currentTimestamp, uint32 incomingTimestamp);
+
+  // Feed object used for storing feed data.
+  // not included but contained in reports:
+  // - blocknumberUpperBound
+  // - upperBlockhash
+  // - blocknumberLowerBound
+  // - currentBlockTimestamp
+  struct Feed {
+    uint32 observationsTimestamp; // the timestamp of the most recent data assigned to this feed
+    int192 price; // the current price of the feed
+    int192 bid; // the current bid price of the feed
+    int192 ask; // the current ask price of the feed
+    string feedName; // the name of the feed
+    string feedId; // the id of the feed (hex encoded)
+    bool active; // true if the feed is being actively updated, otherwise false
+  }
+
+  // Report object obtained from off-chain Mercury server.
+  struct Report {
+    bytes32 feedId; // the feed Id of the report
+    uint32 observationsTimestamp; // the timestamp of when the data was observed
+    int192 price; // the median value of the OCR round
+    int192 bid; // the median bid of the OCR round
+    int192 ask; // the median ask if the OCR round
+    uint64 blocknumberUpperBound; // the highest block observed at the time the report was generated
+    bytes32 upperBlockhash; // the blockhash of the highest block observed
+    uint64 blocknumberLowerBound; // the lowest block observed at the time the report was generated
+    uint64 currentBlockTimestamp; // the timestamp of the highest block observed
+  }
+
+  event FeedUpdated(uint32 observationsTimestamp, int192 price, int192 bid, int192 ask, string feedId);
+
+  uint32 private constant MIN_GAS_FOR_PERFORM = 200_000;
+
+  string constant c_feedParamKey = "feedIdHex"; // for Mercury v0.2 - format by which feeds are identified
+  string constant c_timeParamKey = "blockNumber"; // for Mercury v0.2 - format by which feeds are filtered to be sufficiently recent
+  IVerifierProxy public s_verifier; // for Mercury v0.2 - verifies off-chain reports
+
+  int192 constant scale = 1_000_000; // a scalar used for measuring deviation with precision
+  int192 public s_deviationPercentagePPM; // acceptable deviatoin threshold - 1.5% = 15_000, 100% = 1_000_000, etc..
+  uint32 public s_stalenessSeconds; // acceptable staleness threshold - 60 = 1 minute, 300 = 5 minutes, etc..
+
+  string[] public s_feeds; // list of feed Ids
+  mapping(string => Feed) public s_feedMapping; // mapping of feed Ids to stored feed data
```

Comment:
Just curious but why not make this `mapping(bytes32 => Feed)` since the mercury report gives the feed ID as a `bytes32`? That way no need to convert back/forth from/to string?

Reply:
The revert reason has to take in a string. So either way, you are converting back and forth between the types. The way I have it now was easier I think, but in any case you can't escape the conversion. See the `bytes32ToHexString` helper function's comments.

Reply:
Ah yep just noticed. Sounds good!